### ISSUE TITLE
Missing chains in iptables for active jails

### ISSUE THREAD

**Issue by Brauadler** [MailCleaner/MailCleaner#548]
Since the last update  (#547), only the chains "Input", "Forward", "Output" and a chain "mc-webauth-bl" are displayed in iptables in my instance. Despite active jails in the database (all except the blacklists) and ipsets filled with entries, no other chains are displayed. 

Before the update, there was a separate chain for each jail with the entries from the corresponding ipsets and various incorrect chains (mc-exim-bl-bl etc.) which were removed by the update. I assume that the chains for active jails should not be missing so that the firewall works properly? 

Restarting the server, reloading the firewall and fail2ban did not help either.
![IPtables_250923](https://github.com/MailCleaner/MailCleaner/assets/95625809/eb176269-f2cc-4b85-8658-b850034fd5cb)

**Comment by JohnMertz** [comment]
Interesting... The erroneous 'bl-bl' chains being removed was intention. However, the other `ipset` rules should all be implemented within the INPUT chain. Here's the start of `iptable --list` from our MailCleaner Cloud:

```
target     prot opt source               destination
LOG        all  --  anywhere             anywhere             match-set BLACKLISTNET src LOG level warning
REJECT     all  --  anywhere             anywhere             match-set BLACKLISTNET src reject-with icmp-por
t-unreachable
LOG        all  --  anywhere             anywhere             match-set BLACKLISTIP src LOG level warning
REJECT     all  --  anywhere             anywhere             match-set BLACKLISTIP src reject-with icmp-port
-unreachable
LOG        tcp  --  anywhere             anywhere             tcp dpt:ssh match-set mc-ssh-1y src LOG level w
arning
REJECT     tcp  --  anywhere             anywhere             tcp dpt:ssh match-set mc-ssh-1y src reject-with
 icmp-port-unreachable
LOG        tcp  --  anywhere             anywhere             tcp dpt:smtp match-set mc-exim-1y src LOG level
 warning
REJECT     tcp  --  anywhere             anywhere             tcp dpt:smtp match-set mc-exim-1y src reject-wi
th icmp-port-unreachable
LOG        tcp  --  anywhere             anywhere             multiport dports http,https match-set mc-webaut
h-1y src LOG level warning
REJECT     tcp  --  anywhere             anywhere             multiport dports http,https match-set mc-webaut
h-1y src reject-with icmp-port-unreachable
...
```

Those machines received the update exactly like all other machines should have (unlike my development machines, which is why I did not show them).

Have you received the 1.1.0 version of the python libary as provided in the latest `Updater4MC`? If so, does anything change (or do you see errors) with:

```
/var/mailcleaner/.pyenv/versions/3.7.7/bin/dump_fail2ban_config.py
/usr/mailcleaner/etc/init.d/firewall restart
```

**Comment by Brauadler** [comment]
In my Updater4mc log from 23.09.23 the renewed update 81_fail2ban_ports.update is executed. At first glance, everything seems to be working correct. There is a warning that was displayed during the update of pyenv, but it does not prevent further execution: `warning: cannot open /etc/mailname: permission denied` (4 times in whole process).

Afterwards, mailcleaner-library-1.1.0 is successfully uninstalled and the new version mailcleaner-library-1.1.1 is successfully installed. When generating the Fail2ban config, the following messages appear: 
```
Use of uninitialized value $config{"REGISTERED"} in numeric eq (==) at /usr/mailcleaner/bin/dump_firewall.pl line 85.
Use of uninitialized value $config{"REGISTERED"} in numeric eq (==) at /usr/mailcleaner/bin/dump_firewall.pl line 86. 
DUMPSUCCESSFUL
```

Both commands were executed by the system without error, new .local files are created in the /usr/ directory.
```
/var/mailcleaner/.pyenv/versions/3.7.7/bin/dump_fail2ban_config.py
/usr/mailcleaner/etc/init.d/firewall restart
```
From time to time since the pyenv update there is the following message in the log but with the first update 81 fail2ban still worked: 
`fail2ban.filter [66306]: ERROR Error in FilterPyinotify callback: 'module' object has no attribute '_strptime_time'.`

Which file/script adds the firewall rules? Is it the fail2ban_service.py or the BLACKLIST file created by the dump_firewall? The file BLACKLIST is always empty in my firewall folder. Could it be that the variable $blacklist never becomes **1** to execute the following code (359-274),
   ```
     if ( $blacklist == 1 ) {
                foreach my $period (qw( bl 1d 1w 1m 1y )) {
                        foreach my $f2b (keys(%fail2ban_sets)) {
                                my $ports = $services{$fail2ban_sets{$f2b}}[0];
                                $ports =~ s/[:|]/,/;
                                print BLACKLIST "$iptables -I INPUT -p ".lc($services{$fail2ban_sets{$f2b}}[1])." ".($ports =~ m/,/ ? '-m multiport --dports' : '--dport')." $ports -m set --match-set $f2b-$period src -j REJECT\n";
                                print BLACKLIST "$iptables -I INPUT -p ".lc($services{$fail2ban_sets{$f2b}}[1])." ".($ports =~ m/,/ ? '-m multiport --dports' : '--dport')." $ports -m set --match-set $f2b-$period src -j LOG\n";
                        }
                }
                foreach ( qw( BLACKLISTIP BLACKLISTNET ) ) {
                        print BLACKLIST "$iptables -I INPUT -m set --match-set $_ src -j REJECT\n";
                        print BLACKLIST "$iptables -I INPUT -m set --match-set $_ src -j LOG\n\n";
                }
                chmod 0755, $blacklist_script;
                print START "\n$blacklist_script\n";
        }
```
 as I don't have either of these files?

` my @blacklist_files = ('/usr/mailcleaner/etc/firewall/blacklist.txt', '/usr/mailcleaner/etc/firewall/blacklist_custom.txt');`

**Comment by Brauadler** [comment]
I have added an empty blacklist.txt file in the firewall folder and now the missing entries are added.
Can I delete the chain mc-webauth-bl manually, as this is probably a leftover from between versions and was not removed correctly by the stop script?
```
Chain INPUT (policy DROP)
target     prot opt source               destination
LOG        all  --  anywhere             anywhere             match-set BLACKLISTNET src LOG level warning
REJECT     all  --  anywhere             anywhere             match-set BLACKLISTNET src reject-with icmp-port-unreachable
LOG        all  --  anywhere             anywhere             match-set BLACKLISTIP src LOG level warning
REJECT     all  --  anywhere             anywhere             match-set BLACKLISTIP src reject-with icmp-port-unreachable
LOG        tcp  --  anywhere             anywhere             multiport dports http,https match-set mc-webauth-1y src LOG level warning
REJECT     tcp  --  anywhere             anywhere             multiport dports http,https match-set mc-webauth-1y src reject-with icmp-port-unreachable
LOG        tcp  --  anywhere             anywhere             tcp dpt:ssh match-set mc-ssh-1y src LOG level warning
REJECT     tcp  --  anywhere             anywhere             tcp dpt:ssh match-set mc-ssh-1y src reject-with icmp-port-unreachable
LOG        tcp  --  anywhere             anywhere             tcp dpt:smtp match-set mc-exim-1y src LOG level warning
REJECT     tcp  --  anywhere             anywhere             tcp dpt:smtp match-set mc-exim-1y src reject-with icmp-port-unreachable
LOG        tcp  --  anywhere             anywhere             multiport dports http,https match-set mc-webauth-1m src LOG level warning
REJECT     tcp  --  anywhere             anywhere             multiport dports http,https match-set mc-webauth-1m src reject-with icmp-port-unreachable
LOG        tcp  --  anywhere             anywhere             tcp dpt:ssh match-set mc-ssh-1m src LOG level warning
REJECT     tcp  --  anywhere             anywhere             tcp dpt:ssh match-set mc-ssh-1m src reject-with icmp-port-unreachable
LOG        tcp  --  anywhere             anywhere             tcp dpt:smtp match-set mc-exim-1m src LOG level warning
REJECT     tcp  --  anywhere             anywhere             tcp dpt:smtp match-set mc-exim-1m src reject-with icmp-port-unreachable
LOG        tcp  --  anywhere             anywhere             multiport dports http,https match-set mc-webauth-1w src LOG level warning
REJECT     tcp  --  anywhere             anywhere             multiport dports http,https match-set mc-webauth-1w src reject-with icmp-port-unreachable
LOG        tcp  --  anywhere             anywhere             tcp dpt:ssh match-set mc-ssh-1w src LOG level warning
REJECT     tcp  --  anywhere             anywhere             tcp dpt:ssh match-set mc-ssh-1w src reject-with icmp-port-unreachable
LOG        tcp  --  anywhere             anywhere             tcp dpt:smtp match-set mc-exim-1w src LOG level warning
REJECT     tcp  --  anywhere             anywhere             tcp dpt:smtp match-set mc-exim-1w src reject-with icmp-port-unreachable
LOG        tcp  --  anywhere             anywhere             multiport dports http,https match-set mc-webauth-1d src LOG level warning
REJECT     tcp  --  anywhere             anywhere             multiport dports http,https match-set mc-webauth-1d src reject-with icmp-port-unreachable
LOG        tcp  --  anywhere             anywhere             tcp dpt:ssh match-set mc-ssh-1d src LOG level warning
REJECT     tcp  --  anywhere             anywhere             tcp dpt:ssh match-set mc-ssh-1d src reject-with icmp-port-unreachable
LOG        tcp  --  anywhere             anywhere             tcp dpt:smtp match-set mc-exim-1d src LOG level warning
REJECT     tcp  --  anywhere             anywhere             tcp dpt:smtp match-set mc-exim-1d src reject-with icmp-port-unreachable
LOG        tcp  --  anywhere             anywhere             multiport dports http,https match-set mc-webauth-bl src LOG level warning
REJECT     tcp  --  anywhere             anywhere             multiport dports http,https match-set mc-webauth-bl src reject-with icmp-port-unreachable
LOG        tcp  --  anywhere             anywhere             tcp dpt:ssh match-set mc-ssh-bl src LOG level warning
REJECT     tcp  --  anywhere             anywhere             tcp dpt:ssh match-set mc-ssh-bl src reject-with icmp-port-unreachable
LOG        tcp  --  anywhere             anywhere             tcp dpt:smtp match-set mc-exim-bl src LOG level warning
REJECT     tcp  --  anywhere             anywhere             tcp dpt:smtp match-set mc-exim-bl src reject-with icmp-port-unreachable
...
```
So I assume that it was really due to dump_firewall.pl and the variable.

**Comment by JohnMertz** [comment]
Ah. I understand. Silly mistake and it should be a simple fix to generate the content of `blacklist` regardless of whether `blacklist.txt` exists. Should have it fixed shortly.

Yes you can delete the leftover policy.

**Comment by JohnMertz** [comment]
Should be resolved by #549
